temporal-serviceclient/src/main/java/io/temporal/internal/retryer/GrpcSyncRetryer.java

```diff
@@ -0,0 +1,92 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.retryer;
+
+import io.grpc.StatusRuntimeException;
+import io.temporal.internal.BackoffThrottler;
+import io.temporal.serviceclient.RpcRetryOptions;
+import java.time.Clock;
+import java.util.concurrent.CancellationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class GrpcSyncRetryer {
+  private static final Logger log = LoggerFactory.getLogger(GrpcSyncRetryer.class);
+
+  private final Clock clock;
+
+  public GrpcSyncRetryer(Clock clock) {
+    this.clock = clock;
+  }
+
+  public <R, T extends Throwable> R retry(
+      RpcRetryOptions options, GrpcRetryer.RetryableFunc<R, T> r) throws T {
+    int attempt = 0;
+    long startTime = clock.millis();
+    BackoffThrottler throttler =
+        new BackoffThrottler(
+            options.getInitialInterval(),
+            options.getMaximumInterval(),
+            options.getBackoffCoefficient());
+
+    Exception lastException = null;
+    while (!GrpcRetryerUtils.ranOutOfRetries(options, startTime, attempt, clock.millis())) {
+      attempt++;
+
+      if (lastException != null) {
+        log.warn("Retrying after failure", lastException);
+      }
+
+      try {
+        throttler.throttle();
+        R result = r.apply();
+        throttler.success();
+        return result;
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new CancellationException();
+      } catch (StatusRuntimeException e) {
+        lastException = e;
```

Comment:
Do deadline exceeded exceptions have any cause? If not, should we init the cause and specify previous error there? This way we'll have context of what has caused deadline exceeded error in the first place.

Reply:
I feel a little bit uneasy about it.
First of all, StatusRuntimeException actually doesn't have a cause at all. It doesn't expose any constructor accepting cause. I think it's done for it being serializable.
Second, I feel that such usage of `cause` can be very misleading for users. The previous exception didn't really cause DeadlineExceeded, it was caused by a deadline timeout. 
I understand what you are coming from, but to do that we probably should make a new class of exceptions. Let's catch up on that and make it in a separate PR if we really want it.